eng/ApiListing.targets

```diff
@@ -0,0 +1,53 @@
+<Project>
+
+  <Target Name="GenerateApiListing">
+    <PropertyGroup>
+      <_RefSourceOutputPath>$([System.IO.Directory]::GetParent('$(MSBuildProjectDirectory)'))/ref/</_RefSourceOutputPath>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <_AllTargetFrameworks Remove="@(_AllTargetFrameworks)" />
+      <_AllTargetFrameworks Include="$(TargetFrameworks);$(TargetFramework)" />
+      <_DeduplicatedTargetFramework Include="%(_AllTargetFrameworks.Identity)" />
+    </ItemGroup>
+
+    <MSBuild Projects="$(MSBuildProjectFullPath)"
+             Targets="_GenerateApiListingInner"
+             Properties="TargetFramework=%(_DeduplicatedTargetFramework.Identity)">
+    </MSBuild>
+
+  </Target>
+
+  <Target Name="_GenerateApiListingInner" DependsOnTargets="Build" Condition="'$(GenerateAPIListing)' == 'true'">
```

Comment:
I'm generally not a fan of mutating the source code in the source tree during the build automatically. If we did this I think we should consider generating it to the obj folder and then doing a diff or something similar.

Reply:
It's not during the build, developer has to run eng\Generate-API to update this code. And there is a CI check to verify it's up-to-date.

Reply:
I don't want to slow down build for the rare times public API changes are made. CI checks would remind them :)

Reply:
Where is `eng\Generate-API.ps1`?

Reply:
Renamed to Export-API

Reply:
https://github.com/Azure/azure-sdk-for-net/pull/8695/files#diff-3d8d4efb9238e6d9bd7e0ebe9d325476

Reply:
Or do you mean `Export-API.ps1` and if that is the same then how is this different from the CI?  

We also need some docs updated in Contributing.md to tell folks how to do this.

Reply:
> how is this different from the CI?

It runs a custom target `GenerateAPIListing` that is not executed by default.

Reply:
Where is it failing the build if there is a diff?

Reply:
https://github.com/Azure/azure-sdk-for-net/pull/8695/files#diff-e393debd7f57cca40fe083227f50bc35

https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/templates/jobs/archetype-sdk-client.yml#L91